scipy/special/xsf/tools.h

```diff
@@ -261,10 +258,271 @@ namespace detail {
      * See ContinuedFractionSeriesGenerator for details.
      */
     template <typename Generator, typename T = pair_value_t<generator_result_t<Generator>>>
-    XSF_HOST_DEVICE ContinuedFractionSeriesGenerator<Generator, T>
-    continued_fraction_series(Generator &cf) {
+    XSF_HOST_DEVICE ContinuedFractionSeriesGenerator<Generator, T> continued_fraction_series(Generator &cf) {
         return ContinuedFractionSeriesGenerator<Generator, T>(cf);
     }
 
+    /* Find initial bracket for a bracketing scalar root finder. A valid bracket is a pair of points a < b for
+     * which the signs of f(a) and f(b) differ. Assumes function is monotonic and it is known
+     * whether the function is increasing or decreasing. Passing in whether the function is
+     * increasing or not allows for simplication of logic and avoidance of some branches.
+     *
+     * Note that this takes a pointer to a function taking a tuple of args along with a scalar
+     * double argument. A tuple of args for specializing func is also passed as the final argument.
+     * It would be much cleaner to use std::function and capturing lambda's to specialize the
+     * function we are finding a root for, but this was found unworkable in CuPy using NVRTC.
+     * This should be revisited in the future in order to allow simplifying this code. */
+    template <typename... Args>
+    XSF_HOST_DEVICE inline std::tuple<double, double, int> bracket_root(
+        double (*func)(double, std::tuple<Args...>), double x_left, double x_right, double x_min, double x_max,
+        double factor, bool increasing, std::uint64_t maxiter, std::tuple<Args...> args
+    ) {
```

Comment:
Would it be helpful to validate the input arguments, e.g. `x_min <= x_left < x_right <= x_max`? It could help catch errors such as swapping `x_min` with `x_left` and `x_max` with `x_right`, and would automatically catch `nan` in any of these arguments.